### PR TITLE
fix propagation of channelRegistered event in ChannelInitializer [#4994]

### PR DESCRIPTION
**Pull request by tbcs**
This is a proposed fix for #4994 (please see the issue for further details).

### CODE REVIEW

**Comment by tbcs** [comment]
There was a bug in the test case that I needed to fix (c56b09e).

**Comment by normanmaurer** [comment]
@tbcs can you squash your commits ?

**Comment by tbcs** [comment]
@normanmaurer Sure, I can squash the commits.  Is it OK to do a forced push to the topic branch?

**Comment by normanmaurer** [comment]
@tbcs yes it is :)

**Comment by normanmaurer** [comment]
@tbcs also please adjust the commit message to match our template:

http://netty.io/wiki/writing-a-commit-message.html

**Comment by tbcs** [comment]
@normanmaurer Squashed, commit message adjusted, pushed.

File: transport/src/main/java/io/netty/channel/ChannelInitializer.java [hunk 1]
```diff
@@ -65,9 +64,9 @@
     @Override
     @SuppressWarnings("unchecked")
     public final void channelRegistered(ChannelHandlerContext ctx) throws Exception {
-        initChannel((C) ctx.channel());
         ctx.pipeline().remove(this);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelInitializer.java]: actually I think the remove should still happen after the `initChannel(...)`

File: transport/src/test/java/io/netty/channel/ChannelInitializerTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        final InspectableHandler testSubject = new InspectableHandler();
+        server.childHandler(new ChannelInitializer<LocalChannel>() {
+            @Override
+            public void initChannel(LocalChannel channel) {
+                channel.pipeline().addFirst(testSubject);
+            }
+        });
+        server.bind().sync();
+        client.connect(SERVER_ADDRESS).sync();
+        assertTrue(testSubject.channelRegisteredFired);
```
Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/ChannelInitializerTest.java]: ensure you close the server and client channel.

File: transport/src/test/java/io/netty/channel/ChannelInitializerTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        final InspectableHandler testSubject = new InspectableHandler();
+        server.childHandler(new ChannelInitializer<LocalChannel>() {
+            @Override
+            public void initChannel(LocalChannel channel) {
+                channel.pipeline().addLast(testSubject);
+            }
+        });
+        server.bind().sync();
+        client.connect(SERVER_ADDRESS).sync();
+        assertTrue(testSubject.channelRegisteredFired);
```
Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/ChannelInitializerTest.java]: ensure you close the server and client channel.

File: transport/src/test/java/io/netty/channel/ChannelInitializerTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    }
```
Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/ChannelInitializerTest.java]: Also you can share most of the test-code and just use a different ChannelInitializer for each of them.

**Comment by normanmaurer** [comment]
@tbcs please address my comments and ping me once done.

**Comment by normanmaurer** [comment]
@Scottmitch @trustin @nmittler WDYT about always let the registration event start on the head of the pipeline after initChannel ? I think it makes perfect sense.

**Comment by nmittler** [comment]
@normanmaurer wouldn't this result in handlers receiving the registered event multiple times?

**Comment by normanmaurer** [comment]
@nmittler you mean if `ChannelInitializer` is not the only handler in the pipeline and other handlers exists before it? I think yes this may be possible, I wonder what would be a good way to fix this .

**Comment by nmittler** [comment]
@normanmaurer not sure ... maybe we just accept that it may happen in strange corner cases?

**Comment by normanmaurer** [comment]
@nmittler let me have a cold :beer: and think about ;)

**Comment by nmittler** [comment]
@normanmaurer ha! The source of all good ideas! :)

**Comment by Scottmitch** [comment]
I agree the possibility of multiple calls to channelRegistered will also create confusion, although it may be a less common use case. If `ChannelInitializer` doesn't want to support this use case perhaps we should explicitly check if `ChannelInitializer` is the first in the pipeline when `ChannelHandlerContext` is called. By default this could throw some sort of ISE saying the pipeline configuration is not supported by `ChannelInitializer` (with a comment about how channelRegistered may be called multiple times if allowed), and we `ChannelInitializer` could support an "opt out" mode to allow this behavior. @normanmaurer @nmittler - WDYT?

**Comment by normanmaurer** [comment]
@trustin WDYT ? I'm not sure about this yet.

**Comment by tbcs** [comment]
@normanmaurer I addressed the issues you raised earlier.  I also changed the test handler to count the registeredChannel events instead of just setting a flag, it is probably better to be explicit here.

**Comment by normanmaurer** [comment]
@Scottmitch throwing an exception would be a API breakage... I think I have an idea... stand by :)

**Comment by normanmaurer** [comment]
@trustin btw what is your stance on this ?

**Comment by normanmaurer** [comment]
@trustin ping... at the moment I think we should just use `ChannelPipeline.fireChannelRegistered()` and log a warning if the `ChannelInitializer` is not the first handler in the pipeline. Alternative we may not log anything at all. WDYT ?

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler I'm leaning to just pull this in. I think this will cause the least surprising behaviour at least. WDYT ?

**Comment by nmittler** [comment]
@normanmaurer it's probably fine ... @Scottmitch and @trustin, thoughts?

**Comment by Scottmitch** [comment]
@normanmaurer @nmittler - no strong objection here

**Comment by trustin** [comment]
Fine by me.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (9d4fae308c09a2c759dc67db765d311154c8d977) and 4.0 (4fdf0e8026eab296c24ab41a89f2be16d7b428d9)

@tbcs thanks again!

**Pull request closed without merging**